samcli/commands/deploy/command.py

```diff
@@ -154,27 +169,47 @@ def do_cli(
     role_arn,
     notification_arns,
     fail_on_empty_changeset,
+    use_json,
     tags,
+    metadata,
     region,
     profile,
 ):
+    from samcli.commands.package.package_context import PackageContext
     from samcli.commands.deploy.deploy_context import DeployContext
 
-    with DeployContext(
-        template_file=template_file,
-        stack_name=stack_name,
-        s3_bucket=s3_bucket,
-        force_upload=force_upload,
-        s3_prefix=s3_prefix,
-        kms_key_id=kms_key_id,
-        parameter_overrides=parameter_overrides,
-        capabilities=capabilities,
-        no_execute_changeset=no_execute_changeset,
-        role_arn=role_arn,
-        notification_arns=notification_arns,
-        fail_on_empty_changeset=fail_on_empty_changeset,
-        tags=tags,
-        region=region,
-        profile=profile,
-    ) as deploy_context:
-        deploy_context.run()
+    with tempfile.NamedTemporaryFile() as output_template_file:
+
+        with PackageContext(
+            template_file=template_file,
+            s3_bucket=s3_bucket,
+            s3_prefix=s3_prefix,
+            output_template_file=output_template_file.name,
+            kms_key_id=kms_key_id,
+            use_json=use_json,
+            force_upload=force_upload,
+            metadata=metadata,
+            on_deploy=True,
+            region=region,
+            profile=profile,
+        ) as package_context:
+            package_context.run()
```

Comment:
why are you packaging *always*? This would be a breaking change

Reply:
No. its an attempt to package, if the codeuri's already contain packaged contents, we do not repackage.

Reply:
This is skirting the lines of a breaking change. Is there any other mechanism to create two clear code paths, one for previous experience and other for new one?

And, this relies on the fact that `sam package` is idempotent. That's an undocumented assumption of `sam package`.

Reply:
https://github.com/awslabs/aws-sam-cli/issues/1533. Added this question here.